src/esp/gfx/CubeMap.cpp

```diff
@@ -0,0 +1,438 @@
+    }
+  }
+
+  // depth texture
+  if (flags_ & Flag::DepthTexture) {
+    auto& depthTexture = textures_[TextureType::Depth];
+    depthTexture = std::make_unique<Mn::GL::CubeMapTexture>();
+    (*depthTexture)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setStorage(1, Mn::GL::TextureFormat::DepthComponent32F, size);
+  }
+}
+
+void CubeMap::recreateFramebuffer() {
+  Mn::Vector2i viewportSize{imageSize_, imageSize_};
+  frameBuffer_ = Mn::GL::Framebuffer{{{}, viewportSize}};
+  // optional depth buffer is 24-bit integer pixel, which is different from the
+  // depth texture (32-bit float)
+  optionalDepthBuffer_.setStorage(Mn::GL::RenderbufferFormat::DepthComponent24,
+                                  viewportSize);
+}
+
+void CubeMap::prepareToDraw(unsigned int cubeSideIndex) {
+  mapForDraw(cubeSideIndex);
+
+  // sorry, unlike color buffers, for depth buffer you have to reattach it every
+  // time
+  // however, if NOT using depth texture, we do not need to attach the depth
+  // buffer again and again
+  if (flags_ & Flag::DepthTexture) {
+    Magnum::GL::CubeMapCoordinate cubeMapCoord =
+        convertFaceIndexToCubeMapCoordinate(cubeSideIndex);
+    frameBuffer_.attachCubeMapTexture(
+        Mn::GL::Framebuffer::BufferAttachment::Depth,
+        *textures_[TextureType::Depth], cubeMapCoord, 0);
+  }
+
+  frameBuffer_.clearDepth(1.0f).clearColor(0,                // color attachment
+                                           Mn::Vector4ui{0}  // clear color
+  );
+
+  CORRADE_INTERNAL_ASSERT(
+      frameBuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+}
+
+void CubeMap::mapForDraw(unsigned int index) {
+  frameBuffer_.mapForDraw({
+      {Mn::Shaders::Generic3D::ColorOutput,
+       Mn::GL::Framebuffer::ColorAttachment{index}},
+      // TODO:
+      //{Mn::Shaders::Generic3D::ObjectIdOutput, objectIdAttachment}
+  });
+}
+
+Mn::GL::CubeMapTexture& CubeMap::getTexture(TextureType type) {
+  textureTypeSanityCheck(flags_, type, "CubeMap::getTexture():");
+  return *textures_[type];
+}
+
+#ifndef MAGNUM_TARGET_WEBGL
+// because Mn::Image2D image = textures_[type]->image(...)
+// requires desktop OpenGL
+bool CubeMap::saveTexture(TextureType type,
+                          const std::string& imageFilePrefix) {
+  textureTypeSanityCheck(flags_, type, "CubeMap::saveTexture():");
+
+  Cr::PluginManager::Manager<Mn::Trade::AbstractImageConverter> manager;
+  Cr::Containers::Pointer<Mn::Trade::AbstractImageConverter> converter;
+  if (!(converter = manager.loadAndInstantiate("AnyImageConverter"))) {
+    return false;
+  }
+
+  const char* coordStrings[6] = {"+X", "-X", "+Y", "-Y", "+Z", "-Z"};
+  for (int iFace = 0; iFace < 6; ++iFace) {
+    Mn::Image2D image = textures_[type]->image(
```

Comment:
@mosra :
For depth texture, can I get the image of each side using this `image(...)`?
I I tried to print out the 1st 30 floats (30 depth values) using:
```
        Mn::Debug{} << Cr::Containers::arrayCast<float>(image.data().prefix(120));
```
And they were all 0.

(I also verified the color texture, it was OK. Non-zero values. `saveTexture` and `loadTexture` worked as expected for the rgb cubemap, but not for depth cubemap, as all the pixels were "0".)

Reply:
If you get all zeros, then you probably get a GL error as well (you can check with the `--magnum-gpu-validation` option). Which is probably due to the R32F format being used. If you do the following:

```cpp
        convertFaceIndexToCubeMapCoordinate(iFace), 0, {GL::PixelFormat::DepthComponent, GL::PixelType::Float});
```

it should give you the proper values, I think. But then you have to "reinterpret" the image as a R32F again when saving the image. Something like

```cpp
ImageView2D depthAsRedChannel{image.storage(), PixelFormat::R32F, image.size(), image.data()};
converter->exportToFile(depthAsRedChannel, filename) ...
```

Reply:
quoted from @mosra from other conversations :
"because you're trying to read R32F while it's a depth format (for that you need to use a GL::PixelFormat and GL::PixelType pair, as there are no generic depth PixelFormat values)"
"to explain further: because no image formats so far support storing depth data, there's no generic depth PixelFormat ... but that will probably change with the EXR support implemented, and all this will become simpler"